### PR TITLE
Add opaque yarpc.Headers type

### LINKED ISSUE #68: Expose Headers at the top-level
Users shouldn't have to use `transport.Headers`.

### PR DESCRIPTION
**Pull request by abhinav**
This adds an `internal.Headers` opaque type which backs both, `yarpc.Headers`
and `transport.Headers`. Both of those expose the subsets of the API they care
about.

A feature of the new Headers representation is that zero-value is always valid.
You can do,

```go
var headers yarpc.Headers  // or transport.Headers or internal.Headers
headers = headers.With("a", "1").With("b", "2")
headers = headers.With("c", "3")
```

Or,

```go
headers := yarpc.NewHeaders().With("a", "1")
headers = headers.With("b", "2")
```

Note that the Headers object returned by `With` **must** be used in place of
the original because it *may* not point to the same underlying map.

`yarpc.Headers`, `transport.Headers` and `internal.Headers`, all expose the
following functions.

    func (h Headers) Del(k string)
    func (h Headers) Get(k string) (string, bool)
    func (h Headers) Len() int
    func (h Headers) With(k, v string) Headers

`yarpc.Headers` exposes the following additional functions.

    func NewHeaders() Headers
    func (h Headers) Keys() []string

`transport.Headers` exposes the following additional functions:

    func HeadersFromMap(m map[string]string) Headers
    func NewHeaders() Headers
    func NewHeadersWithCapacity(capacity int) Headers
    func (h Headers) Items() map[string]string

Resolves #68, #212

-------------------------------------------------------------------------------

Note: This diff modifies a large number of files but the bulk of those changes
are mechanical updates of `transport.Headers{...}`, to
`transport.NewHeaders().With(..).With(..)`. I recommend reviewing individual
commits rather than the whole diff. In particular, the fist 5 commits.

### CODE REVIEW

**Comment by abhinav** [comment]
Something seems to be wrong on Travis' Go 1.5 here. It's failing to run `go test` but everything else is working just fine. Locally, Go 1.5 works just fine for me.

**Comment by abhinav** [comment]
@yarpc/golang This is now ready to review. Reviewing individual commits still recommended.

File: header.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+//
+// The returned Headers object MAY use a different underlying data store than
+// the original Headers object, and so the returned object MUST always be used
+// instead of the original object.
+//
+// 	headers = headers.With("foo", "bar")
+//
+// This call may be chained to set multiple headers consecutively.
+//
+// 	headers = headers.With("foo", "bar").With("baz", "qux")
```
Comment (breerly) [inline header.go]: Any reason not to put an empty comment line after this to separate the example from the implementation?

File: header.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+// Len returns the number of headers defined on this object.
+func (h Headers) Len() int {
+	return internal.Headers(h).Len()
+}
+
+// Keys returns a list of header keys defined on this Headers object.
+//
+// All items in the list will be normalized using CanonicalizeHeaderKey.
+// The returned list is for reading only. Mutating the list will result in
+// undefined behavior.
```
Comment (breerly) [inline header.go]: "undefined behavior"... for example?

File: internal/headers.go [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+import "strings"
+
+var emptyMap = map[string]string{}
+
+// Headers provides the implementation of both, yarpc.Headers and
+// transport.Headers.
+//
+// Keys in the map are canonicalized using CanonicalizeHeaderKey.
+type Headers struct {
```
Comment (breerly) [inline internal/headers.go]: Any reason this can't be just `type Headers map[string]string`?

File: internal/headers.go [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+		h.items = make(map[string]string)
+	}
+	h.items[CanonicalizeHeaderKey(k)] = v
+	return h
+}
+
+// Items returns the underlying map for this Headers object. Mutation of this
+// map will result in undefined behavior.
+//
+// This ALWAYS returns a non-nil result.
```
Comment (breerly) [inline internal/headers.go]: +1 ty sir

File: internal/headers.go [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+func (h Headers) With(k, v string) Headers {
+	if h.items == nil {
+		h.items = make(map[string]string)
+	}
+	h.items[CanonicalizeHeaderKey(k)] = v
+	return h
+}
+
+// Items returns the underlying map for this Headers object. Mutation of this
+// map will result in undefined behavior.
```
Comment (breerly) [inline internal/headers.go]: Again not sure what you mean by undefined behavior here...

File: internal/headers.go [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+func (h Headers) Items() map[string]string {
+	if h.items == nil {
+		return emptyMap
+	}
+	return h.items
+}
+
+// Get retrieves the value associated with the given key.
+func (h Headers) Get(k string) (string, bool) {
+	v, ok := h.items[CanonicalizeHeaderKey(k)]
```
Comment (breerly) [inline internal/headers.go]: Perhaps we should normalize on write instead of read.

File: internal/headers.go [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+	return v, ok
+}
+
+// Del deletes the header with the given name.
+func (h Headers) Del(k string) {
+	delete(h.items, CanonicalizeHeaderKey(k))
```
Comment (breerly) [inline internal/headers.go]: Instead of duplicating the normalization everywhere you read, suggest calling `Get` from here, or creating a new function that both `Get` and `Del` call.

File: header.go (hunk 2 above)
Comment (abhinav) [inline header.go]: Undefined behavior as in the behavior is not defined. As in, don't do this because we don't guarantee what will happen if you do it and you can't rely on it.

Although for `yarpc.Headers` since we'll probably almost always allocate the slice, we can just remove this line and let the users do whatever with it.

File: internal/headers.go (hunk 3 above)
Comment (abhinav) [inline internal/headers.go]: Because we want the type to be opaque to the outside world. Making this `map[string]string` means that both, `yarpc.Headers` and `transport.Headers` (which are aliases of this) can be instantiated with invalid values using `yarpc.Headers{"Foo": "bar"}`.

File: internal/headers.go (hunk 7 above)
Comment (breerly) [inline internal/headers.go]: Actually, I was taking for granted that the term `Canonicalize` is commonly used in-place of normalize. 

Deleting this comment :)

File: internal/headers.go (hunk 6 above)
Comment (abhinav) [inline internal/headers.go]: We normalize on write too. We need to normalize on read because the written version is normalized.

```go
h = h.With("Foo", "bar") // stores "foo": "bar"
h.Get("Foo")  // needs to normalize "Foo" to "foo" to match the stored value
```

File: transport/header.go [hunk 8]
```diff
@@ -20,41 +20,76 @@
...
+}
+
+// Headers is the transport-level representation of application headers.
 //
-// Header keys are case insensitive and duplicates are disallowed.
-type Headers map[string]string
+// Keys in the map MUST be canonicalized with CanonicalizeHeaderKey.
+//
+// You probably want to look at yarpc.Headers instead.
+type Headers internal.Headers
```
Comment (breerly) [inline transport/header.go]: I'm not clear why we need to re-expose the header func down in the transport package. Is the `yarpc` and `internal` package not enough?

File: internal/headers.go (hunk 5 above)
Comment (prashantv) [inline internal/headers.go]: I would specifically call out that the resulting map should be treated as read only and MUST not be mutated -- mainly since in the case we return the `emptyMap`, mutation does very bad things

File: internal/headers.go [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+}
+
+// Len returns the number of headers defined on this object.
+func (h Headers) Len() int {
+	return len(h.items)
+}
+
+// CanonicalizeHeaderKey canonicalizes the given header key for storage into
+// the Headers map.
+func CanonicalizeHeaderKey(k string) string {
```
Comment (prashantv) [inline internal/headers.go]: TODO for dealing with unsupported characters in HTTP headers keys?

File: transport/header.go (hunk 8 above)
Comment (abhinav) [inline transport/header.go]: Explained offline

File: transport/header.go [hunk 10]
```diff
@@ -20,41 +20,76 @@
...
+//
+// Mutating the returned map will result in undefined behavior.
+func (h Headers) Items() map[string]string {
+	return internal.Headers(h).Items()
+}
+
+// HeadersFromMap builds a new Headers object from the given map of header
+// key-value pairs.
+func HeadersFromMap(m map[string]string) Headers {
+	if m == nil {
```
Comment (prashantv) [inline transport/header.go]: nit: do `if len(m) == 0`

File: transport/tchannel/header.go [hunk 11]
```diff
@@ -36,14 +36,14 @@ import (
 
 // pullBaggage pulls the context headers from the given transport.Headers,
 // deleting them from the original headers map.
-func pullBaggage(headers transport.Headers) transport.Headers {
-	ctxHeaders := make(transport.Headers)
+func pullBaggage(headers transport.Headers) map[string]string {
```
Comment (breerly) [inline transport/tchannel/header.go]: Can you explain why you're using `map[string]string` here (and in several other places) instead of a defined type?

File: internal/headers.go (hunk 3 above)
Comment (breerly) [inline internal/headers.go]: Aaah, yeah, `items` is private. +1

File: header.go [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+func NewHeaders() Headers {
+	return Headers{}
+}
+
+// With returns a Headers object with the given key-value pair added to it.
+// If a header with the same name already exists, it will be overwritten.
+//
+// The returned Headers object MAY use a different underlying data store than
+// the original Headers object, and so the returned object MUST always be used
+// instead of the original object.
```
Comment (prashantv) [inline header.go]: maybe call out that this is similar to `append`? Don't want users thinking that `With` will return a new Headers always, since the old value may be modified and can't be reused

File: internal/headers.go (hunk 6 above)
Comment (breerly) [inline internal/headers.go]: +1

File: header.go [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+func (h Headers) Del(k string) {
+	internal.Headers(h).Del(k)
+}
+
+// Len returns the number of headers defined on this object.
+func (h Headers) Len() int {
+	return internal.Headers(h).Len()
+}
+
+// Keys returns a list of header keys defined on this Headers object.
```
Comment (prashantv) [inline header.go]: should this method sort the keys so that it's deterministic  and consistent? Otherwise, call out that the keys are in an undefined order (which is a little annoying to use)

**Comment by breerly** [comment]
lgtm - 🐑  it

File: transport/transporttest/context.go [hunk 14]
```diff
@@ -103,7 +104,7 @@ func (c *ContextMatcher) Matches(got interface{}) bool {
 
 	if c.baggage != nil {
 		headers := baggage.FromContext(ctx)
-		if !reflect.DeepEqual(c.baggage, headers) {
+		if !reflect.DeepEqual(*c.baggage, headers) {
```
Comment (prashantv) [inline transport/transporttest/context.go]: can baggage ever be `nil`? If so, do:
```go
if !reflect.DeepEqual(c.baggage, &headers) {
```

if it should never be `nil`, this is fine, as it'll cause a panic and we'll notice very quickly

File: crossdock/server/yarpc/echo_test.go [hunk 15]
```diff
@@ -1,67 +0,0 @@
-package yarpc
```
Comment (prashantv) [inline crossdock/server/yarpc/echo_test.go]: why is this file being deleted in this change?

**Comment by prashantv** [comment]
Commits say `Use new transport.Headers interface` -- don't think we have an interface for `transport.Headers` anymore.

Finished doing a review pass -- left comments inside of the commits

File: crossdock/server/yarpc/echo_test.go (hunk 15 above)
Comment (breerly) [inline crossdock/server/yarpc/echo_test.go]: We shouldn't be unit testing things crossdock already tests.

Although to @prashantv's point, that probably should go in another PR.

File: transport/transporttest/context.go (hunk 14 above)
Comment (abhinav) [inline transport/transporttest/context.go]: `c.baggage`? There's a `!=` check above.

File: header.go (hunk 1 above)
Comment (abhinav) [inline header.go]: 👍

File: internal/headers.go (hunk 5 above)
Comment (abhinav) [inline internal/headers.go]: 👍

File: internal/headers.go (hunk 9 above)
Comment (abhinav) [inline internal/headers.go]: Added TODO

File: transport/header.go (hunk 10 above)
Comment (abhinav) [inline transport/header.go]: 👍

File: transport/tchannel/header.go (hunk 11 above)
Comment (abhinav) [inline transport/tchannel/header.go]: This returns a `map[string]string` because `baggage.NewContextWithHeaders` expects one. That expects a map because it's just an internal bulk version of `baggage.NewContext(ctx, string, string)`. And since it has to construct a new Headers object regardless, it will end up normalizing all the keys again anyway.

File: header.go (hunk 12 above)
Comment (abhinav) [inline header.go]: 👍

File: header.go (hunk 13 above)
Comment (abhinav) [inline header.go]: Good catch. Sorting.

File: crossdock/server/yarpc/echo_test.go (hunk 15 above)
Comment (abhinav) [inline crossdock/server/yarpc/echo_test.go]: True. Extracting to https://github.com/yarpc/yarpc-go/pull/219

**Comment by abhinav** [comment]
> Commits say Use new transport.Headers interface -- don't think we have an interface for transport.Headers anymore.

I meant interface as in API. I'll reword the commits.

**Comment by abhinav** [comment]
Updated. Ready for another pass.

**Comment by prashantv** [comment]
lgtm
